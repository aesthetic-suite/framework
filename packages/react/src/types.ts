// Any must be used here for consumers to be typed correctly.
/* eslint-disable @typescript-eslint/no-explicit-any */

import React from 'react';
import { Theme, ClassNameSheetVariants } from '@aesthetic/core';
import { ClassName, Direction, ThemeName } from '@aesthetic/types';

export type ClassNameGenerator<T> = (
  ...keys: (undefined | null | false | T | ClassNameSheetVariants)[]
) => ClassName;

// CONTEXT

export type DirectionContextType = Direction;

export interface DirectionProviderProps {
  children: NonNullable<React.ReactNode>;
  /** Explicit direction to use. */
  direction?: Exclude<Direction, 'neutral'>;
  /** Render an inline element instead of block. */
  inline?: boolean;
  /** Locale aware string to deduce the direction from. */
  value?: string;
}

export type ThemeContextType = Theme;

export interface ThemeProviderProps {
  children: NonNullable<React.ReactNode>;
  /** Theme to activate. */
  name?: ThemeName;
}

// HOCs

export interface WrapperComponent {
  WrappedComponent: React.ComponentType<any>;
}

export interface WrapperProps {
  /** Gain a reference to the wrapped component. */
  wrappedRef?: React.Ref<any>;
}

export interface WrappedProps {
  /** The ref passed by the `wrappedRef` prop. */
  ref?: React.Ref<any>;
}

export interface WithDirectionWrappedProps extends WrappedProps {
  /** The current direction. Provided by `withDirection`. */
  direction: Direction;
}

export interface WithThemeWrappedProps extends WrappedProps {
  /** The theme object. Provided by `withTheme`. */
  theme: Theme;
}

export interface WithStylesWrappedProps<T = string> extends WrappedProps {
  /** Function to generate CSS class names from a style sheet. Provided by `withStyles`. */
  cx: ClassNameGenerator<T>;
}
