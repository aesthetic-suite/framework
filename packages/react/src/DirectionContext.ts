import React from 'react';
import { getActiveDirection } from '@aesthetic/core';
import { DirectionContextType } from './types';

export default React.createContext<DirectionContextType>(getActiveDirection() || 'ltr');
