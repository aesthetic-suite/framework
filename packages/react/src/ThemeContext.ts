import React from 'react';
import { ThemeContextType } from './types';

export default React.createContext<ThemeContextType | null>(null);
