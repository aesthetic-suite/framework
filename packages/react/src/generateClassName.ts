import { ClassNameSheet } from '@aesthetic/core';
import { ClassName } from '@aesthetic/types';
import { arrayLoop, isObject, objectLoop } from '@aesthetic/utils';

export default function generateClassName<T extends unknown[]>(
  keys: T,
  classNames: ClassNameSheet<string>,
): ClassName {
  const list: string[] = [];

  arrayLoop(keys, (key) => {
    // Variants
    if (isObject(key)) {
      objectLoop(key, (value, type) => {
        if (!value || !type) {
          return;
        }

        const variant = `${type}_${value}`;

        objectLoop(classNames, (clx) => {
          if (clx?.variants?.[variant]) {
            list.push(clx.variants[variant]);
          }
        });
      });

      // Rules/styles
    } else if (typeof key === 'string' && classNames[key]?.class) {
      list.push(classNames[key]!.class);
    }
  });

  return list.join(' ');
}
