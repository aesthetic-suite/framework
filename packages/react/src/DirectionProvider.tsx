import React from 'react';
import getDirection from 'direction';
import { Direction } from '@aesthetic/types';
import DirectionContext from './DirectionContext';
import { DirectionProviderProps } from './types';

/**
 * Explicitly define a direction or automatically infer a direction based on a string of text.
 * Will render an element with a `dir` attribute set.
 */
export default function DirectionProvider({
  children,
  direction,
  inline,
  value,
}: DirectionProviderProps) {
  const Tag = inline ? 'span' : 'div';
  // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
  const dir: Direction = direction || getDirection(value) || 'ltr';

  return (
    <DirectionContext.Provider value={dir}>
      <Tag dir={dir}>{children}</Tag>
    </DirectionContext.Provider>
  );
}
