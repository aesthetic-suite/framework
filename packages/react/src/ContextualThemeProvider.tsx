import React, { useState, useEffect } from 'react';
import { getTheme, renderThemeStyles } from '@aesthetic/core';
import ThemeContext from './ThemeContext';
import useDirection from './useDirection';
import { ThemeProviderProps } from './types';

/**
 * Rendered nested within the page to provide a contextual theme.
 */
export default function ContextualThemeProvider({ children, name }: Required<ThemeProviderProps>) {
  const [className, setClassName] = useState('');
  const direction = useDirection();
  const theme = getTheme(name);

  // Render styles when theme/direction change
  useEffect(() => {
    setClassName(renderThemeStyles(theme, { direction }));
  }, [theme, direction]);

  return (
    <ThemeContext.Provider value={theme}>
      <div className={className} data-theme={name}>
        {children}
      </div>
    </ThemeContext.Provider>
  );
}
