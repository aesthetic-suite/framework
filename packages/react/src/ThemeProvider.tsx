import React, { useState, useEffect, useCallback } from 'react';
import { changeTheme, getActiveTheme, getTheme, subscribe, unsubscribe } from '@aesthetic/core';
import ThemeContext from './ThemeContext';
import { ThemeProviderProps } from './types';

/**
 * Rendered at the root to provide the theme to the entire application.
 */
export default function ThemeProvider({ children, name = '' }: ThemeProviderProps) {
  const [themeName, setThemeName] = useState(name);
  const theme = themeName ? getTheme(themeName) : getActiveTheme();

  // Listen to theme changes that occur outside of the provider
  const handleChangeTheme = useCallback(
    (nextTheme) => {
      if (nextTheme !== themeName) {
        setThemeName(nextTheme);
      }
    },
    [themeName],
  );

  useEffect(() => {
    subscribe('change:direction', handleChangeTheme);
    subscribe('change:theme', handleChangeTheme);

    return () => {
      unsubscribe('change:direction', handleChangeTheme);
      unsubscribe('change:theme', handleChangeTheme);
    };
  }, [handleChangeTheme]);

  // Update state when the `name` prop changes
  useEffect(() => {
    if (name) {
      changeTheme(name, false);
      setThemeName(name);
    }
  }, [name]);

  return <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>;
}
