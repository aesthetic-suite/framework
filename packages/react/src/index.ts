/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import DirectionContext from './DirectionContext';
import DirectionProvider from './DirectionProvider';
import ThemeContext from './ThemeContext';
import ThemeProvider from './ThemeProvider';
import ContextualThemeProvider from './ContextualThemeProvider';
import useDirection from './useDirection';
import useStyles from './useStyles';
import useTheme from './useTheme';
import withDirection from './withDirection';
import withStyles from './withStyles';
import withTheme from './withTheme';

export * from '@aesthetic/core';
export * from './types';

export {
  DirectionContext,
  DirectionProvider,
  ThemeContext,
  ThemeProvider,
  ContextualThemeProvider,
  useDirection,
  useStyles,
  useTheme,
  withDirection,
  withStyles,
  withTheme,
};
