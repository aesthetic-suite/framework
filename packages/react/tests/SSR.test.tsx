/* eslint-disable react/jsx-no-literals */

import React from 'react';
import { renderToString, renderToStaticMarkup } from 'react-dom/server';
import { aesthetic, LocalSheet, SheetStructure } from '@aesthetic/core';
import { ServerRenderer } from '@aesthetic/style/server';
import { setupAesthetic, teardownAesthetic, purgeStyles } from '@aesthetic/core/src/testing';
import { useStyles, ThemeProvider } from '../src';
import { createStyleSheet, ButtonProps } from './__mocks__/Button';

describe('SSR', () => {
  let renderer: ServerRenderer;
  let sheet: LocalSheet<SheetStructure<
    'button' | 'button_block' | 'button_disabled' | 'button_large' | 'button_small'
  >>;

  function Button({ children, block, disabled, large, small }: ButtonProps) {
    const cx = useStyles(sheet);

    return (
      <button
        type="button"
        className={cx('button', block && 'button_block', disabled && 'button_disabled', {
          // eslint-disable-next-line no-nested-ternary
          size: large ? 'lg' : small ? 'sm' : 'df',
        })}
      >
        {children}
      </button>
    );
  }

  function App() {
    return (
      <ThemeProvider name="night">
        <main>
          <div>You are not logged in!</div>
          <Button href="/login">Login</Button>
          <Button href="/register">Register</Button>
        </main>
      </ThemeProvider>
    );
  }

  beforeEach(() => {
    renderer = new ServerRenderer();
    sheet = createStyleSheet();

    setupAesthetic(aesthetic);
  });

  afterEach(() => {
    teardownAesthetic(aesthetic);
    purgeStyles();

    // @ts-expect-error
    delete global.AESTHETIC_CUSTOM_RENDERER;
  });

  describe('renderToString()', () => {
    it('renders markup', () => {
      expect(renderToString(renderer.extractStyles(<App />))).toMatchSnapshot();
      expect(renderer.renderToStyleMarkup()).toMatchSnapshot();
    });
  });

  describe('renderToStaticMarkup()', () => {
    it('renders markup', () => {
      expect(renderToStaticMarkup(renderer.extractStyles(<App />))).toMatchSnapshot();
      expect(renderer.renderToStyleMarkup()).toMatchSnapshot();
    });
  });
});
