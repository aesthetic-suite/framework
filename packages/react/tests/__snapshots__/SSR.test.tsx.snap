// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSR renderToStaticMarkup() renders markup 1`] = `"<main><div>You are not logged in!</div><button type=\\"button\\" class=\\"p q r s t u v w x y z a1 b1 c1 d1 e1 f1 g1 h1 i1 j1 k1 l1 m1 n1 o1 p1 f g h i j\\">Login</button><button type=\\"button\\" class=\\"p q r s t u v w x y z a1 b1 c1 d1 e1 f1 g1 h1 i1 j1 k1 l1 m1 n1 o1 p1 f g h i j\\">Register</button></main>"`;

exports[`SSR renderToStaticMarkup() renders markup 2`] = `"<style id=\\"aesthetic-global\\" type=\\"text/css\\" media=\\"screen\\" data-aesthetic-type=\\"global\\" data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-index=\\"51\\"></style><style id=\\"aesthetic-standard\\" type=\\"text/css\\" media=\\"screen\\" data-aesthetic-type=\\"standard\\" data-aesthetic-hydrate-index=\\"51\\" data-aesthetic-rule-index=\\"51\\">.a { min-width:7.50rem; }.b { padding-top:var(--spacing-sm); }.c { padding-bottom:var(--spacing-sm); }.d { padding-left:var(--spacing-df); }.e { padding-right:var(--spacing-df); }.f { min-width:10rem; }.g { padding-top:var(--spacing-df); }.h { padding-bottom:var(--spacing-df); }.i { padding-left:var(--spacing-md); }.j { padding-right:var(--spacing-md); }.k { min-width:12.50rem; }.l { padding-top:var(--spacing-md); }.m { padding-bottom:var(--spacing-md); }.n { padding-left:var(--spacing-lg); }.o { padding-right:var(--spacing-lg); }.p { appearance:none; }.q { background-color:var(--palette-brand-color-40); }.r { border:0; }.s { cursor:pointer; }.t { display:inline-flex; }.u { font-size:inherit; }.v { margin:0; }.w { padding:0; }.x { text-decoration:none; }.y { user-select:auto; }.z { vertical-align:middle; }.a1 { border-color:var(--palette-neutral-color-30); }.b1 { border-radius:var(--border-df-radius); }.c1 { border-style:solid; }.d1 { border-width:var(--border-df-width); }.e1 { position:relative; }.f1 { text-align:center; }.g1 { color:var(--palette-neutral-color-00); }.h1::-moz-focus-inner { border:0; }.i1::-moz-focus-inner { padding:0; }.j1::-moz-focus-inner { margin:0; }.k1:focus { border-color:var(--palette-primary-bg-focused); }.l1:hover { border-color:var(--palette-primary-bg-hovered); }.m1:hover { background-color:var(--palette-brand-color-50); }.n1:active { border-color:var(--palette-primary-bg-selected); }.o1:active { background-color:var(--palette-brand-color-60); }.p1[disabled] { border-color:var(--palette-primary-bg-disabled); }.q1 { overflow-wrap:normal; }.r1 { word-wrap:normal; }.s1 { word-break:normal; }.t1 { display:block; }.u1 { width:100%; }.v1 { white-space:normal; }.w1 { overflow:hidden; }.x1 { opacity:0.4; }.y1 { opacity:0.5; }.z1 { opacity:0.6; }</style><style id=\\"aesthetic-conditions\\" type=\\"text/css\\" media=\\"screen\\" data-aesthetic-type=\\"conditions\\" data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-index=\\"51\\"></style>"`;

exports[`SSR renderToString() renders markup 1`] = `"<main><div>You are not logged in!</div><button type=\\"button\\" class=\\"p q r s t u v w x y z a1 b1 c1 d1 e1 f1 g1 h1 i1 j1 k1 l1 m1 n1 o1 p1 f g h i j\\">Login</button><button type=\\"button\\" class=\\"p q r s t u v w x y z a1 b1 c1 d1 e1 f1 g1 h1 i1 j1 k1 l1 m1 n1 o1 p1 f g h i j\\">Register</button></main>"`;

exports[`SSR renderToString() renders markup 2`] = `"<style id=\\"aesthetic-global\\" type=\\"text/css\\" media=\\"screen\\" data-aesthetic-type=\\"global\\" data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-index=\\"51\\"></style><style id=\\"aesthetic-standard\\" type=\\"text/css\\" media=\\"screen\\" data-aesthetic-type=\\"standard\\" data-aesthetic-hydrate-index=\\"51\\" data-aesthetic-rule-index=\\"51\\">.a { min-width:7.50rem; }.b { padding-top:var(--spacing-sm); }.c { padding-bottom:var(--spacing-sm); }.d { padding-left:var(--spacing-df); }.e { padding-right:var(--spacing-df); }.f { min-width:10rem; }.g { padding-top:var(--spacing-df); }.h { padding-bottom:var(--spacing-df); }.i { padding-left:var(--spacing-md); }.j { padding-right:var(--spacing-md); }.k { min-width:12.50rem; }.l { padding-top:var(--spacing-md); }.m { padding-bottom:var(--spacing-md); }.n { padding-left:var(--spacing-lg); }.o { padding-right:var(--spacing-lg); }.p { appearance:none; }.q { background-color:var(--palette-brand-color-40); }.r { border:0; }.s { cursor:pointer; }.t { display:inline-flex; }.u { font-size:inherit; }.v { margin:0; }.w { padding:0; }.x { text-decoration:none; }.y { user-select:auto; }.z { vertical-align:middle; }.a1 { border-color:var(--palette-neutral-color-30); }.b1 { border-radius:var(--border-df-radius); }.c1 { border-style:solid; }.d1 { border-width:var(--border-df-width); }.e1 { position:relative; }.f1 { text-align:center; }.g1 { color:var(--palette-neutral-color-00); }.h1::-moz-focus-inner { border:0; }.i1::-moz-focus-inner { padding:0; }.j1::-moz-focus-inner { margin:0; }.k1:focus { border-color:var(--palette-primary-bg-focused); }.l1:hover { border-color:var(--palette-primary-bg-hovered); }.m1:hover { background-color:var(--palette-brand-color-50); }.n1:active { border-color:var(--palette-primary-bg-selected); }.o1:active { background-color:var(--palette-brand-color-60); }.p1[disabled] { border-color:var(--palette-primary-bg-disabled); }.q1 { overflow-wrap:normal; }.r1 { word-wrap:normal; }.s1 { word-break:normal; }.t1 { display:block; }.u1 { width:100%; }.v1 { white-space:normal; }.w1 { overflow:hidden; }.x1 { opacity:0.4; }.y1 { opacity:0.5; }.z1 { opacity:0.6; }</style><style id=\\"aesthetic-conditions\\" type=\\"text/css\\" media=\\"screen\\" data-aesthetic-type=\\"conditions\\" data-aesthetic-hydrate-index=\\"-1\\" data-aesthetic-rule-index=\\"51\\"></style>"`;
